### DISCUSSION TITLE
OBS Hardware Survey

### DISCUSSION THREAD

**Discussion by derrod** [obsproject/obs-studio#6282] [General]
# OBS Hardware Survey

The idea of doing an "OBS Hardware Survey" inspired by the [Steam Hardware Survey](https://store.steampowered.com/hwsurvey) has been floated for a while. This discussion is meant to have a more permanent location to keep some notes on the idea and benefits/concerns around it.

## Motivation

We do not have any telemetry (or plans for any) in OBS. The hardware survey would provide us with potentially valuable information on how and on what platforms our users use OBS.  
The survey would be strictly opt-in (just like Steam's survey) and offered periodically to users. 

This would be useful for a few things:
- Inform test setups for common configurations (hardware/OBS setup)
- Helping with prioritisation decisions with regards to support/development for various platforms and hardware
- Better judging of the impact of issues with certain configurations or development decisions

## Collection process

1. When OBS queries the server for updates, it also queries whether a hardware survey round is currently enabled
    1. The survey may only be enabled for a subset of users, this could be decided based on the existing random and anonymous installation GUID
2. If enabled, the user would be prompted to ask if they want to participate in the survey with a short summary of what it does, with a "no" and "next" button
    1.  If accepted, the next page would contain a text box disclosing the exact information collected, with a "cancel" and "next" option
    2. the third page shows a progress bar or loading spinner while data is collected
    3. The last page thanks the user for participating
3. A background process collects the system data and submits it to the obsproject servers
4. The data is processed and stored in a database

The results can then be published to the public on the obsproject website after the data collection round is over.

## Data collected

The data collected would largely be the same as presented in logs, although more machine-readable.

- OBS version/runtime
- OS version/edition (language?)
- Anti-virus software?
- Hardware
    + Device type (notebook, desktop, etc.)
    + CPU (features, cores, etc.)
    + GPU(s) (primary + secondary, encoder capabilities)
    + RAM (amount, + currently used by OBS)
    + Display(s) (number, resolutions, capabilities, etc.) 
    + Input/Output devices used (sound card, webcams, capture cards, etc.)
- Loaded (third-party) plugins
- OBS configuration meta-information (number of profiles/scene collections, etc.)
- Current OBS setup (anonymised scene tree, integration usage, service used, encoder/graphics configuration)
- Network connectivity (wifi or wired, perhaps auto-config bandwidth results?)
- Potentially information extracted from existing logs (profiler info? crashes? did adaptive bitrate kick in?)

<details>
  <summary>Anonymised scene tree example</summary>

```json
{
  "1": [
    {
      "type": "directshow",
      "enabled": true,
      "id": "4"
    },
    {
      "type": "browser",
      "enabled": false,
      "id": "5"
    },
    {
      "type": "group",
      "enabled": true,
      "children": [
        {
          "type": "text",
          "enabled": false
        }
      ],
      "id": "6"
    }
  ],
  "2": [
    {
      "type": "browser",
      "enabled": false,
      "id": "5"
    },
    {
      "type": "ffmpeg_source",
      "enabled": true,
      "id": "7"
    }
  ],
  "3": [
    {
      "type": "scene",
      "enabled": true,
      "target": "1",
      "id": "8"
    }
  ]
}
```

</details>

## Technical considerations

- Survey collection should work on all supported platforms (Windows, macOS, Linux) and distribution methods (Direct, Steam, Flathub, etc.)
- Some protection against submitting fraudulent survey data may be necessary (rate limits, plausibility checking)
- The survey may run as a separate application to allow for it to be updated independently of OBS in case the collected data needs to be adjusted
- Data needs to be stored in a manner that can be processed in a reasonable amount of time to produce or update reports

## Privacy considerations

- Surveys would be submitted without any personal information attached, meaning no paths, user information, etc.
- Data may be manually reviewed by obsproject members before publication to avoid disclosing information about unreleased or unannounced hardware/software
    + Uncommon or unknown devices could be filtered out automatically, or potentially limited to manufacturer name/vendor IDs
    + Alternatively this data could be submitted in a hashed form, known items (and thus known hashes) could be unhashed without revealing information about unknown items

**Comment by Fenrirthviti** [comment]
Possibly out of scope, but I wouldn't mind throwing in a few basic questions based on feature use/voting kinda things.

**Reply by derrod** [reply]
I think that's something more geared towards a general user survey, since it requires user-interaction past just clicking "yes" I don't tthink it's really fit for this specific idea.
